python/cudf/cudf/core/index.py

```diff
@@ -2744,6 +2748,212 @@ def categories(self):
+            closed='left',
+            dtype='interval')
+
+    """
+    if freq and periods and start and end:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, exactly three must be specified"
+        )
+    elif periods and not freq:
+        assert end is not None and start is not None
+        end = end + 1
+        periods_array = cupy.asarray(cupy.arange(start, end))
+        _, bin_edges = cupy.histogram(periods_array, periods)
+        # cupy.histogram turns all arrays into a float array
+        # this can cause the dtype to be a float instead of an int
+        # the below adjusts for this
+        if cupy.all(cupy.mod(bin_edges, 1) == 0):
+            bin_edges = bin_edges.astype(int)
+        left_col = bin_edges[:-1]
+        right_col = bin_edges[1:]
+    elif freq and periods:
+        if end:
+            start = end - (freq * periods)
+        if start:
+            end = freq * periods + start
+        left_col = cupy.arange(start, end, freq)
+        assert end is not None and start is not None
+        end = end + 1
+        start = start + freq
+        right_col = cupy.arange(start, end, freq)
+    elif freq and not periods:
+        assert end is not None and start is not None
+        end = end - freq + 1
+        left_col = cupy.arange(start, end, freq)
+        end = end + freq + 1
+        start = start + freq
+        right_col = cupy.arange(start, end, freq)
+    elif start is not None and end is not None:
+        left_col = cupy.arange(start, end, freq)
+        start = start + 1
+        end = end + 1
+        right_col = cupy.arange(start, end, freq)
+    else:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, atleast two must be specified"
+        )
+    if len(right_col) == 0 or len(left_col) == 0:
+        return cudf.IntervalIndex([], closed=closed)
+
+    interval_col = column.build_interval_column(
+        left_col, right_col, closed=closed
+    )
+    return IntervalIndex(interval_col)
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    IntervalIndex.
+
+    closed : {"left", "right", "both", "neither"}, default "right"
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    dtype : dtype or None, default None
+    If None, dtype will be inferred.
+
+    copy : bool, default False
+    Copy the input data.
+
+    name : object, optional
+    Name to be stored in the index.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy(deep=True)
```

Comment:
It's a shame that if we're constructing from a CPU object here that we copy it to the GPU and then deep copy it unnecessarily.

Reply:
Removing the extra copies 😄